Basic framework for Distributed Autograd context.

pritamdamania87 opened the pull request:
Summary:
As per https://github.com/pytorch/pytorch/issues/23110, each autograd pass
would be assigned a unique autograd_context_id. In this change we introduce a
DistAutogradContainer per worker which holds information for each autograd pass
currently running.

DistAutogradContainer has a map from the autograd_context_id to
DistAutogradContext (which holds all the relevant information for the autograd
pass). DistAutogradContext currently only stores the autograd_context_id and
more information would be added to it later as we build out the rest of the
framework.

The autograd_context_id is a 64 bit globally unique integer where the first 16
bits are the worker_id and next 48 bits are auto-incrementing for uniqueness.

Sample python code on how this would be used for distributed autograd:

```
import torch.distributed.autograd as dist_autograd
worker_id = 0
dist_autograd.init(worker_id)
with dist_autograd.context() as context_id:
     # forward pass...
     # backward pass...
     # optimizer step...
```

Differential Revision: D16356694

mrshenli commented on test/test_dist_autograd.py:
```diff
@@ -0,0 +1,20 @@
+
+    def test_autograd_context(self):
+        worker_id = 16
+        dist_autograd.init(worker_id)
```
if `dist_autograd` always work with `RPC` (I assume so, as it needs send/recv), can we merge this with `init_rpc`, so that users don't have to call two init API? For example, it could be `init_rpc` by default also initialize dist autograd, but there could an arg to disable it if the application won't run any backward on top of RPC.

mrshenli commented on torch/csrc/distributed/autograd/context/DistAutogradContainer.h:
```diff
@@ -0,0 +1,59 @@
+  int64_t current_context_id_;
+
+  // Unique id to identify a worker in the distributed setting.
+  int16_t worker_id_;
```
If using ProcessGroup, will this be the rank? I believe rank is `int` in `ProcessGroup`, any reason for truncate it to 16 bits?

ezyang commented:
> DistAutogradContainer has a map from the autograd_context_id to
DistAutogradContext (which holds all the relevant information for the autograd
pass).

How does garbage collection (e.g., not leaking memory) work if you have a map? (Haven't looked at the code yet, but from reading the PR description this is an obvious question I have.)

ezyang commented on torch/distributed/autograd/__init__.py:
```diff
@@ -0,0 +1,25 @@
+if hasattr(torch._C, "_dist_autograd_init") and not torch._C._dist_autograd_init():
+    raise RuntimeError("Failed to initialize PyTorch distributed autograd support")
+
+class context:
```
If you're going to omit `context(object)`, it should be a lot more obvious in this file that it is Python 3 only...

ezyang commented on torch/distributed/autograd/__init__.py:
```diff
@@ -0,0 +1,25 @@
+class context:
+    '''
+        Autograd context object to wrap forward and backward passes when using
+        distributed autograd.
```
Since this API isn't needed in normal PyTorch usage, it's good to briefly explain why this is needed (and what the consequences are of getting this wrong). Also, in the docs here, you say "forward and backward passes", but then in the example below you include the optimizer step. Should the optimizer step be included?

ezyang commented on test/test_dist_autograd.py:
```diff
@@ -0,0 +1,20 @@
+        with dist_autograd.context() as context_id:
+            self.assertEqual(context_id, dist_autograd.retrieve_context(context_id).context_id())
+            # First 16 bits should be worker_id.
+            self.assertEqual(worker_id, context_id >> 48)
```
Is this actually documented behavior that we are committing to permanently, or do we actually just want the ID to be an opaque integer that we have full control over?

ezyang commented on torch/csrc/distributed/autograd/init.cpp:
```diff
@@ -0,0 +1,66 @@
+
+  module.def("retrieve_context", [](int64_t context_id) {
+    return DistAutogradContainer::getInstance().retrieveContext(context_id);
+  });
```
These need docs. Because they are bound from C++, the current idiom is to bind docstrings to these methods from `torch.distributed.autograd`

ezyang replied:
Hmm, but in this method definition style (where you import a module and attach methods to it) you might have an ordering problem. In any case, we need docs on these methods somehow...

ezyang replied:
If I understand correctly, this Python binding style was copied from the existing c10d code. That's understandable. I'm not a big fan of this binding style though, because it means in a file like `torch/distributed/autograd/__init__.py` you call `release_context` and I'm like, "What the heck, where did this identifier come from?" Essentially, you are monkeypatching these definitions into the module.

A more ordinary and explicable way of exposing modules is to actually define a C binding module (similar to `torch._C`) and then reference this C exposed module explicitly.

ezyang replied:
It also makes writing type stubs easier, since you no longer have to write a type stub for all of `torch/distributed/autograd/__init__.py`, you can just write a type stub for the `_C` module, and then put types inline in `torch/distributed/autograd/__init__.py` (because you're Python 3 only and you can do that, right? :)

pritamdamania87 replied:
Thanks for the suggestion, I'll look into what we do for `torch._C` and update the PR accordingly.

pritamdamania87 replied:
Looking at torch._C it seems like that binding code (https://github.com/pytorch/pytorch/blob/c5482e33e9ef59af6c18059378964a00036d95aa/torch/csrc/autograd/python_variable.cpp#L538-L621) is much more complicated than what we have here.

ezyang commented on torch/csrc/distributed/autograd/context/DistAutogradContainer.cpp:
```diff
@@ -0,0 +1,60 @@
+  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  autograd_context_.emplace(
+      current_context_id_, DistAutogradContext(current_context_id_));
+  return autograd_context_.at(current_context_id_++);
```
nit: Let's check for overflow into the worker_id portion. It's unlikely to happen but if it does, the check will save you a lot of heartache.

ezyang commented on torch/csrc/distributed/autograd/context/DistAutogradContainer.h:
```diff
@@ -0,0 +1,59 @@
+  bool initialized_;
+
+  // Lock to protect current_context_id_ and autograd_context map.
+  std::mutex autograd_context_lock_;
```
The other fields are immutable? Worth saying so explicitly.

pritamdamania committed:
Basic framework for Distributed Autograd context. (#24020)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/24020

As per https://github.com/pytorch/pytorch/issues/23110, each autograd pass
would be assigned a unique autograd_context_id. In this change we introduce a
DistAutogradContainer per worker which holds information for each autograd pass
currently running.

DistAutogradContainer has a map from the autograd_context_id to
DistAutogradContext (which holds all the relevant information for the autograd
pass). DistAutogradContext currently only stores the autograd_context_id and
more information would be added to it later as we build out the rest of the
framework.

The autograd_context_id is a 64 bit globally unique integer where the first 16
bits are the worker_id and next 48 bits are auto-incrementing for uniqueness.

Sample python code on how this would be used for distributed autograd:

```
import torch.distributed.autograd as dist_autograd
worker_id = 0
dist_autograd.init(worker_id)
with dist_autograd.context() as context_id:
     # forward pass...
     # backward pass...
     # optimizer step...
```

Differential Revision: D16356694

fbshipit-source-id: 0c2e93b22d892e3220a1197cb544e4d8f8c0990a

mrshenli commented on torch/distributed/__init__.py:
```diff
@@ -19,4 +20,28 @@ def is_available():
+                Arguments:
+                    worker_name (str): a globally unique name of this node. (e.g.,
+                                       ``Trainer3``, ``ParameterServer2``, ``Master``, ``Worker1``)
+                    worker_id (int): a globally unique id for this node. Typically, can use the
```
does this have to be specified by users, can the backend instances figure this out on their own? Users wouldn't need this id any more this point, so it seems like this should be a purely internal variable?

pritamdamania87 replied:
This is possible for the ProcessGroup backend, but I'm not sure how the thrift backend would figure this out automatically. Also, note that even for the ProcessGroup backend the user does have to specify this when they initialize the ProcessGroup.

pritamdamania87 commented:
Abandoning this in favor of https://github.com/pytorch/pytorch/pull/24875 since I had to rework a few things based on ghstack.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/DistAutogradContainer.cpp:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  if (current_context_id_ == std::numeric_limits<int64_t>::max() ||
```
In https://github.com/pytorch/pytorch/pull/24876, I use -1 to indicate that we don't have a valid current autograd context.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/DistAutogradContainer.cpp:
```diff
@@ -0,0 +1,75 @@
+
+  std::lock_guard<std::mutex> guard(autograd_context_lock_);
+  if (current_context_id_ == std::numeric_limits<int64_t>::max() ||
+      current_context_id_ >
```
`current_context_id_` is entirely managed by this class and there shouldn't be a way to set this to a negative number externally. Also, this check precisely avoids `current_context_id_` from overflowing into a negative number.

pritamdamania87 commented on torch/csrc/distributed/autograd/context/DistAutogradContainer.cpp:
```diff
@@ -0,0 +1,75 @@
+  if (current_context_id_ == std::numeric_limits<int64_t>::max() ||
+      current_context_id_ >
+          (kMaxContextId |
+           (static_cast<int64_t>(worker_id_) << kContextIdBits))) {
```
The logic here would become simpler, but we would have something similar to this when we assemble the two. We're essentially moving the bit operations from one place to another.
